File: src/libqimagewriter.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qimageiohandler_enums = @import("libqimageiohandler.zig").enums;
const qimagewriter_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qimagewriter.html
pub const qimagewriter = struct {
    /// New constructs a new QImageWriter object.
    ///
    ///
    pub fn New() QtC.QImageWriter {
        return qtc.QImageWriter_new();
    }

    /// New2 constructs a new QImageWriter object.
    ///
    /// ``` device: QtC.QIODevice, format: []u8 ```
    pub fn New2(device: ?*anyopaque, format: []u8) QtC.QImageWriter {
        const format_str = qtc.libqt_string{
            .len = format.len,
            .data = format.ptr,
        };

        return qtc.QImageWriter_new2(@ptrCast(device), format_str);
    }

    /// New3 constructs a new QImageWriter object.
    ///
    /// ``` fileName: []const u8 ```
    pub fn New3(fileName: []const u8) QtC.QImageWriter {
        const fileName_str = qtc.libqt_string{
            .len = fileName.len,
            .data = fileName.ptr,
        };

        return qtc.QImageWriter_new3(fileName_str);
    }

    /// New4 constructs a new QImageWriter object.
    ///
    /// ``` fileName: []const u8, format: []u8 ```
    pub fn New4(fileName: []const u8, format: []u8) QtC.QImageWriter {
        const fileName_str = qtc.libqt_string{
            .len = fileName.len,
            .data = fileName.ptr,
        };
        const format_str = qtc.libqt_string{
            .len = format.len,
            .data = format.ptr,
        };

        return qtc.QImageWriter_new4(fileName_str, format_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qobject.html#tr)
    ///
    /// ``` sourceText: []const u8, allocator: std.mem.Allocator ```
    pub fn Tr(sourceText: []const u8, allocator: std.mem.Allocator) []const u8 {
        const sourceText_Cstring = sourceText.ptr;
        const _str = qtc.QImageWriter_Tr(sourceText_Cstring);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qimagewriter.Tr: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setFormat)
    ///
    /// ``` self: QtC.QImageWriter, format: []u8 ```
    pub fn SetFormat(self: ?*anyopaque, format: []u8) void {
        const format_str = qtc.libqt_string{
            .len = format.len,
            .data = format.ptr,
        };
        qtc.QImageWriter_SetFormat(@ptrCast(self), format_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#format)
    ///
    /// ``` self: QtC.QImageWriter, allocator: std.mem.Allocator ```
    pub fn Format(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QImageWriter_Format(@ptrCast(self));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qimagewriter.Format: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setDevice)
    ///
    /// ``` self: QtC.QImageWriter, device: QtC.QIODevice ```
    pub fn SetDevice(self: ?*anyopaque, device: ?*anyopaque) void {
        qtc.QImageWriter_SetDevice(@ptrCast(self), @ptrCast(device));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#device)
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn Device(self: ?*anyopaque) QtC.QIODevice {
        return qtc.QImageWriter_Device(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setFileName)
    ///
    /// ``` self: QtC.QImageWriter, fileName: []const u8 ```
    pub fn SetFileName(self: ?*anyopaque, fileName: []const u8) void {
        const fileName_str = qtc.libqt_string{
            .len = fileName.len,
            .data = fileName.ptr,
        };
        qtc.QImageWriter_SetFileName(@ptrCast(self), fileName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#fileName)
    ///
    /// ``` self: QtC.QImageWriter, allocator: std.mem.Allocator ```
    pub fn FileName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QImageWriter_FileName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qimagewriter.FileName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setQuality)
    ///
    /// ``` self: QtC.QImageWriter, quality: i32 ```
    pub fn SetQuality(self: ?*anyopaque, quality: i32) void {
        qtc.QImageWriter_SetQuality(@ptrCast(self), @intCast(quality));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#quality)
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn Quality(self: ?*anyopaque) i32 {
        return qtc.QImageWriter_Quality(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setCompression)
    ///
    /// ``` self: QtC.QImageWriter, compression: i32 ```
    pub fn SetCompression(self: ?*anyopaque, compression: i32) void {
        qtc.QImageWriter_SetCompression(@ptrCast(self), @intCast(compression));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#compression)
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn Compression(self: ?*anyopaque) i32 {
        return qtc.QImageWriter_Compression(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setSubType)
    ///
    /// ``` self: QtC.QImageWriter, typeVal: []u8 ```
    pub fn SetSubType(self: ?*anyopaque, typeVal: []u8) void {
        const typeVal_str = qtc.libqt_string{
            .len = typeVal.len,
            .data = typeVal.ptr,
        };
        qtc.QImageWriter_SetSubType(@ptrCast(self), typeVal_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#subType)
    ///
    /// ``` self: QtC.QImageWriter, allocator: std.mem.Allocator ```
    pub fn SubType(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QImageWriter_SubType(@ptrCast(self));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qimagewriter.SubType: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#supportedSubTypes)
    ///
    /// ``` self: QtC.QImageWriter, allocator: std.mem.Allocator ```
    pub fn SupportedSubTypes(self: ?*anyopaque, allocator: std.mem.Allocator) [][]u8 {
        const _arr: qtc.libqt_list = qtc.QImageWriter_SupportedSubTypes(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]u8, _arr.len) catch @panic("qimagewriter.SupportedSubTypes: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("qimagewriter.SupportedSubTypes: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setOptimizedWrite)
    ///
    /// ``` self: QtC.QImageWriter, optimize: bool ```
    pub fn SetOptimizedWrite(self: ?*anyopaque, optimize: bool) void {
        qtc.QImageWriter_SetOptimizedWrite(@ptrCast(self), optimize);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#optimizedWrite)
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn OptimizedWrite(self: ?*anyopaque) bool {
        return qtc.QImageWriter_OptimizedWrite(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setProgressiveScanWrite)
    ///
    /// ``` self: QtC.QImageWriter, progressive: bool ```
    pub fn SetProgressiveScanWrite(self: ?*anyopaque, progressive: bool) void {
        qtc.QImageWriter_SetProgressiveScanWrite(@ptrCast(self), progressive);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#progressiveScanWrite)
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn ProgressiveScanWrite(self: ?*anyopaque) bool {
        return qtc.QImageWriter_ProgressiveScanWrite(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#transformation)
    ///
    /// ``` self: QtC.QImageWriter ```
    ///
    /// Returns: ``` flag of qimageiohandler_enums.Transformation ```
    pub fn Transformation(self: ?*anyopaque) i64 {
        return qtc.QImageWriter_Transformation(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setTransformation)
    ///
    /// ``` self: QtC.QImageWriter, orientation: flag of qimageiohandler_enums.Transformation ```
    pub fn SetTransformation(self: ?*anyopaque, orientation: i64) void {
        qtc.QImageWriter_SetTransformation(@ptrCast(self), @intCast(orientation));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#setText)
    ///
    /// ``` self: QtC.QImageWriter, key: []const u8, text: []const u8 ```
    pub fn SetText(self: ?*anyopaque, key: []const u8, text: []const u8) void {
        const key_str = qtc.libqt_string{
            .len = key.len,
            .data = key.ptr,
        };
        const text_str = qtc.libqt_string{
            .len = text.len,
            .data = text.ptr,
        };
        qtc.QImageWriter_SetText(@ptrCast(self), key_str, text_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#canWrite)
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn CanWrite(self: ?*anyopaque) bool {
        return qtc.QImageWriter_CanWrite(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#write)
    ///
    /// ``` self: QtC.QImageWriter, image: QtC.QImage ```
    pub fn Write(self: ?*anyopaque, image: ?*anyopaque) bool {
        return qtc.QImageWriter_Write(@ptrCast(self), @ptrCast(image));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#error)
    ///
    /// ``` self: QtC.QImageWriter ```
    ///
    /// Returns: ``` qimagewriter_enums.ImageWriterError ```
    pub fn Error(self: ?*anyopaque) i64 {
        return qtc.QImageWriter_Error(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#errorString)
    ///
    /// ``` self: QtC.QImageWriter, allocator: std.mem.Allocator ```
    pub fn ErrorString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QImageWriter_ErrorString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qimagewriter.ErrorString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#supportsOption)
    ///
    /// ``` self: QtC.QImageWriter, option: qimageiohandler_enums.ImageOption ```
    pub fn SupportsOption(self: ?*anyopaque, option: i64) bool {
        return qtc.QImageWriter_SupportsOption(@ptrCast(self), @intCast(option));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#supportedImageFormats)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn SupportedImageFormats(allocator: std.mem.Allocator) [][]u8 {
        const _arr: qtc.libqt_list = qtc.QImageWriter_SupportedImageFormats();
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]u8, _arr.len) catch @panic("qimagewriter.SupportedImageFormats: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("qimagewriter.SupportedImageFormats: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#supportedMimeTypes)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn SupportedMimeTypes(allocator: std.mem.Allocator) [][]u8 {
        const _arr: qtc.libqt_list = qtc.QImageWriter_SupportedMimeTypes();
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]u8, _arr.len) catch @panic("qimagewriter.SupportedMimeTypes: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("qimagewriter.SupportedMimeTypes: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#imageFormatsForMimeType)
    ///
    /// ``` mimeType: []u8, allocator: std.mem.Allocator ```
    pub fn ImageFormatsForMimeType(mimeType: []u8, allocator: std.mem.Allocator) [][]u8 {
        const mimeType_str = qtc.libqt_string{
            .len = mimeType.len,
            .data = mimeType.ptr,
        };
        const _arr: qtc.libqt_list = qtc.QImageWriter_ImageFormatsForMimeType(mimeType_str);
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]u8, _arr.len) catch @panic("qimagewriter.ImageFormatsForMimeType: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("qimagewriter.ImageFormatsForMimeType: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qobject.html#tr)
    ///
    /// ``` sourceText: []const u8, disambiguation: []const u8, allocator: std.mem.Allocator ```
    pub fn Tr2(sourceText: []const u8, disambiguation: []const u8, allocator: std.mem.Allocator) []const u8 {
        const sourceText_Cstring = sourceText.ptr;
        const disambiguation_Cstring = disambiguation.ptr;
        const _str = qtc.QImageWriter_Tr2(sourceText_Cstring, disambiguation_Cstring);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qimagewriter.Tr2: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qobject.html#tr)
    ///
    /// ``` sourceText: []const u8, disambiguation: []const u8, n: i32, allocator: std.mem.Allocator ```
    pub fn Tr3(sourceText: []const u8, disambiguation: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const sourceText_Cstring = sourceText.ptr;
        const disambiguation_Cstring = disambiguation.ptr;
        const _str = qtc.QImageWriter_Tr3(sourceText_Cstring, disambiguation_Cstring, @intCast(n));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("qimagewriter.Tr3: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qimagewriter.html#dtor.QImageWriter)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QImageWriter ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QImageWriter_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qimagewriter.html#types
pub const enums = struct {
    pub const ImageWriterError = enum {
        pub const UnknownError: i32 = 0;
        pub const DeviceError: i32 = 1;
        pub const UnsupportedFormatError: i32 = 2;
        pub const InvalidImageError: i32 = 3;
    };
};

File: src/libqimagewriter.cpp
#include <QByteArray>
#include <QIODevice>
#include <QImage>
#include <QImageWriter>
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qimagewriter.h>
#include "libqimagewriter.h"
#include "libqimagewriter.hxx"

QImageWriter* QImageWriter_new() {
    return new QImageWriter();
}

QImageWriter* QImageWriter_new2(QIODevice* device, const libqt_string format) {
    QByteArray format_QByteArray(format.data, format.len);
    return new QImageWriter(device, format_QByteArray);
}

QImageWriter* QImageWriter_new3(const libqt_string fileName) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    return new QImageWriter(fileName_QString);
}

QImageWriter* QImageWriter_new4(const libqt_string fileName, const libqt_string format) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    QByteArray format_QByteArray(format.data, format.len);
    return new QImageWriter(fileName_QString, format_QByteArray);
}

libqt_string QImageWriter_Tr(const char* sourceText) {
    QString _ret = QImageWriter::tr(sourceText);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QImageWriter_SetFormat(QImageWriter* self, const libqt_string format) {
    QByteArray format_QByteArray(format.data, format.len);
    self->setFormat(format_QByteArray);
}

libqt_string QImageWriter_Format(const QImageWriter* self) {
    QByteArray _qb = self->format();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QImageWriter_SetDevice(QImageWriter* self, QIODevice* device) {
    self->setDevice(device);
}

QIODevice* QImageWriter_Device(const QImageWriter* self) {
    return self->device();
}

void QImageWriter_SetFileName(QImageWriter* self, const libqt_string fileName) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    self->setFileName(fileName_QString);
}

libqt_string QImageWriter_FileName(const QImageWriter* self) {
    QString _ret = self->fileName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QImageWriter_SetQuality(QImageWriter* self, int quality) {
    self->setQuality(static_cast<int>(quality));
}

int QImageWriter_Quality(const QImageWriter* self) {
    return self->quality();
}

void QImageWriter_SetCompression(QImageWriter* self, int compression) {
    self->setCompression(static_cast<int>(compression));
}

int QImageWriter_Compression(const QImageWriter* self) {
    return self->compression();
}

void QImageWriter_SetSubType(QImageWriter* self, const libqt_string typeVal) {
    QByteArray typeVal_QByteArray(typeVal.data, typeVal.len);
    self->setSubType(typeVal_QByteArray);
}

libqt_string QImageWriter_SubType(const QImageWriter* self) {
    QByteArray _qb = self->subType();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of libqt_string */ QImageWriter_SupportedSubTypes(const QImageWriter* self) {
    QList<QByteArray> _ret = self->supportedSubTypes();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * _ret.size()));
    for (size_t i = 0; i < _ret.size(); ++i) {
        QByteArray _lv_qb = _ret[i];
        libqt_string _lv_str;
        _lv_str.len = _lv_qb.length();
        _lv_str.data = static_cast<const char*>(malloc((_lv_str.len + 1) * sizeof(char)));
        memcpy((void*)_lv_str.data, _lv_qb.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void QImageWriter_SetOptimizedWrite(QImageWriter* self, bool optimize) {
    self->setOptimizedWrite(optimize);
}

bool QImageWriter_OptimizedWrite(const QImageWriter* self) {
    return self->optimizedWrite();
}

void QImageWriter_SetProgressiveScanWrite(QImageWriter* self, bool progressive) {
    self->setProgressiveScanWrite(progressive);
}

bool QImageWriter_ProgressiveScanWrite(const QImageWriter* self) {
    return self->progressiveScanWrite();
}

int QImageWriter_Transformation(const QImageWriter* self) {
    return static_cast<int>(self->transformation());
}

void QImageWriter_SetTransformation(QImageWriter* self, int orientation) {
    self->setTransformation(static_cast<QImageIOHandler::Transformations>(orientation));
}

void QImageWriter_SetText(QImageWriter* self, const libqt_string key, const libqt_string text) {
    QString key_QString = QString::fromUtf8(key.data, key.len);
    QString text_QString = QString::fromUtf8(text.data, text.len);
    self->setText(key_QString, text_QString);
}

bool QImageWriter_CanWrite(const QImageWriter* self) {
    return self->canWrite();
}

bool QImageWriter_Write(QImageWriter* self, const QImage* image) {
    return self->write(*image);
}

int QImageWriter_Error(const QImageWriter* self) {
    return static_cast<int>(self->error());
}

libqt_string QImageWriter_ErrorString(const QImageWriter* self) {
    QString _ret = self->errorString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool QImageWriter_SupportsOption(const QImageWriter* self, int option) {
    return self->supportsOption(static_cast<QImageIOHandler::ImageOption>(option));
}

libqt_list /* of libqt_string */ QImageWriter_SupportedImageFormats() {
    QList<QByteArray> _ret = QImageWriter::supportedImageFormats();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * _ret.size()));
    for (size_t i = 0; i < _ret.size(); ++i) {
        QByteArray _lv_qb = _ret[i];
        libqt_string _lv_str;
        _lv_str.len = _lv_qb.length();
        _lv_str.data = static_cast<const char*>(malloc((_lv_str.len + 1) * sizeof(char)));
        memcpy((void*)_lv_str.data, _lv_qb.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of libqt_string */ QImageWriter_SupportedMimeTypes() {
    QList<QByteArray> _ret = QImageWriter::supportedMimeTypes();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * _ret.size()));
    for (size_t i = 0; i < _ret.size(); ++i) {
        QByteArray _lv_qb = _ret[i];
        libqt_string _lv_str;
        _lv_str.len = _lv_qb.length();
        _lv_str.data = static_cast<const char*>(malloc((_lv_str.len + 1) * sizeof(char)));
        memcpy((void*)_lv_str.data, _lv_qb.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_list /* of libqt_string */ QImageWriter_ImageFormatsForMimeType(const libqt_string mimeType) {
    QByteArray mimeType_QByteArray(mimeType.data, mimeType.len);
    QList<QByteArray> _ret = QImageWriter::imageFormatsForMimeType(mimeType_QByteArray);
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * _ret.size()));
    for (size_t i = 0; i < _ret.size(); ++i) {
        QByteArray _lv_qb = _ret[i];
        libqt_string _lv_str;
        _lv_str.len = _lv_qb.length();
        _lv_str.data = static_cast<const char*>(malloc((_lv_str.len + 1) * sizeof(char)));
        memcpy((void*)_lv_str.data, _lv_qb.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string QImageWriter_Tr2(const char* sourceText, const char* disambiguation) {
    QString _ret = QImageWriter::tr(sourceText, disambiguation);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string QImageWriter_Tr3(const char* sourceText, const char* disambiguation, int n) {
    QString _ret = QImageWriter::tr(sourceText, disambiguation, static_cast<int>(n));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc((_str.len + 1) * sizeof(char)));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QImageWriter_Delete(QImageWriter* self) {
    delete self;
}
